ouroboros-consensus/src/Ouroboros/Storage/ChainDB/Impl/Reader.hs

```diff
@@ -0,0 +1,500 @@
+      -> SlotNo
+      -> [Point blk]
+      -> m (Maybe (Point blk))
+    findFirstPointOnChain curChain slotNoAtImmDBTip = \case
+      []     -> return Nothing
+      pt:pts
+        | AF.withinFragmentBounds (castPoint pt) curChain
+        -> do
+          -- It's in the in-memory chain fragment
+          atomically $ writeTVar varReader $ ReaderInMem $ RollBackTo pt
+          return $ Just pt
+
+        | pointSlot pt > slotNoAtImmDBTip
+          -- The point is after the tip of the ImmutableDB, so don't look in
+          -- the ImmutableDB and skip the point.
+        -> findFirstPointOnChain curChain slotNoAtImmDBTip pts
+
+        | pt == genesisPoint
+          -- We know the point is not in the in-memory fragment (not even as
+          -- the anchor point). This means that the ImmutableDB is not empty
+          -- (@AF.anchorPoint curChain /= genesisPoint@).
+        -> do
+          immIt <- ImmDB.streamBlocksAfter cdbImmDB genesisPoint
+          let readerState = ReaderInImmDB (RollBackTo pt) immIt
+          atomically $ writeTVar varReader readerState
+          return $ Just pt
+
+        | otherwise
+        -> do
+          inImmDB <- isJust <$> ImmDB.getBlockWithPoint cdbImmDB pt
+          if inImmDB
+            then do
+              -- TODO combine block checking with opening the iterator to
+              -- avoid reading the same block twice
+              immIt <- ImmDB.streamBlocksAfter cdbImmDB pt
+              let readerState = ReaderInImmDB (RollBackTo pt) immIt
+              atomically $ writeTVar varReader readerState
+              return $ Just pt
+            else findFirstPointOnChain curChain slotNoAtImmDBTip pts
+
+-- | Update the given 'ReaderState' to account for switching the current
+-- chain to the given fork (which might just be an extension of the
+-- current chain).
+--
+-- PRECONDITION: the intersection point must be within the fragment bounds
+-- of the new chain
+switchFork
+  :: forall m blk.
+     HasHeader (Header blk)
+  => Point blk  -- ^ Intersection point between old and new chain
+  -> AnchoredFragment (Header blk)  -- ^ The new chain
+  -> ReaderState m blk -> ReaderState m blk
+switchFork ipoint newChain readerState =
+    assert (AF.withinFragmentBounds (castPoint ipoint) newChain) $
+      case readerState of
+        -- If the reader is still reading from the ImmutableDB, switching to a
+        -- fork won't affect it at all
+        ReaderInImmDB {} -> readerState
+        ReaderInMem   rollState
+          | AF.withinFragmentBounds (currentPoint rollState) newChain
+            -- If the current reader point is still on the new chain fragment,
+            -- we can keep on rolling forward
+          -> readerState
+          | pointSlot ipoint > pointSlot (currentPoint rollState)
+            -- If the intersection point is more recent than the current
+            -- reader point, we can keep on rolling forward
+          -> readerState
+          | otherwise
```

Comment:
Add `AF.withinFragmentBounds (currentPoint rollState) newChain` as an assertion (after inverting the condition in the case above).